File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-21.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = set()
        l, r = 0, 0
        longest = 0

        while r < len(s):
            while s[r] in res:
                res.remove(s[l])
                l += 1
            res.add(s[r])
            longest = max(longest, r - l + 1)
            r += 1
        return longest

File: Data Structures & Algorithms/trapping-rain-water/submission-21.cs
public class Solution {
    public int Trap(int[] height) {
        int l = 0, r = height.Length - 1;
        int maxL = height[l], maxR = height[r];
        int res = 0;

        while (l < r) {
            if (height[l] <= height[r]) {
                l++;
                maxL = Math.Max(maxL, height[l]);
                res += maxL - height[l];
            } else {
                r--;
                maxR = Math.Max(maxR, height[r]);
                res += maxR - height[r];
            }
        }
        return res;
    }
}
